proxy/src/registry/transaction.rs

```diff
@@ -127,6 +202,60 @@ where
                 .expect("unable to get 'transactions' bucket"),
         }
     }
+
+    /// Starts up the machinery to check for latest block information and advance the state of
+    /// cached transactions.
+    ///
+    /// # Errors
+    ///
+    /// Will return `Err` if a protocol error occurs or access to [`kv::Store`] fails.
+    pub async fn run(&self) -> Result<(), error::Error> {
+        let mut interval = tokio::time::interval(Duration::from_secs(10));
+        loop {
+            interval.tick().await;
+
+            self.advance(self.client.latest_height().await?)?;
+        }
+    }
+
+    /// Updates cached transactions progress given the latest height.
+    fn advance(&self, latest_height: protocol::BlockNumber) -> Result<(), error::Error> {
+        let mut txs = self.list_transactions(vec![])?;
+
+        for tx in &mut txs {
+            match tx.state {
+                State::Confirmed {
+                    block, timestamp, ..
+                } => {
+                    let target = block
+                        .checked_add(SETTLEMENT_OFFSET)
+                        .unwrap_or(SETTLEMENT_OFFSET);
+
+                    if latest_height >= target {
+                        tx.state = State::Settled {
+                            timestamp: Timestamp::now(),
+                        };
+                    } else {
+                        let offset = latest_height
+                            .checked_add(SETTLEMENT_OFFSET)
+                            .unwrap_or(SETTLEMENT_OFFSET);
+                        let progress = offset.saturating_sub(target);
+
+                        tx.state = State::Confirmed {
+                            block,
+                            progress,
+                            timestamp,
+                        };
+                    }
+
+                    self.cache_transaction(tx.clone())?;
+                },
+                State::Failed { .. } | State::Pending { .. } | State::Settled { .. } => continue,
```

Comment:
Don't we want to do something if it failed?

Reply:
This routine is about updating the progress (soon renamed to number of confirmations), if tx is already in a failed state there is nothing to update.

Reply:
Ah ok, so you get the information about failure from somewhere else I suppose?

Reply:
Yeah it's in the initial transaction submission. We haven't figured out a way yet to determine if a tx is not part of the best chain anymore.